config/src/main/java/org/springframework/security/config/annotation/web/configuration/OAuth2ClientConfiguration.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.config.annotation.web.configuration;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.ImportSelector;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.web.method.annotation.OAuth2ClientArgumentResolver;
+import org.springframework.util.ClassUtils;
+import org.springframework.web.method.support.HandlerMethodArgumentResolver;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+import java.util.List;
+
+/**
+ * {@link Configuration} for OAuth 2.0 Client support.
+ *
+ * @author Joe Grandja
+ * @since 5.1
+ */
+@Import(OAuth2ClientConfiguration.OAuth2ClientWebMvcImportSelector.class)
+final class OAuth2ClientConfiguration {
+
+	static class OAuth2ClientWebMvcImportSelector implements ImportSelector {
+
+		@Override
+		public String[] selectImports(AnnotationMetadata importingClassMetadata) {
+			boolean webmvcPresent = ClassUtils.isPresent(
+				"org.springframework.web.servlet.DispatcherServlet", getClass().getClassLoader());
+
+			return webmvcPresent ?
+				new String[] { "org.springframework.security.config.annotation.web.configuration.OAuth2ClientConfiguration.OAuth2ClientWebMvcSecurityConfiguration" } :
+				new String[] {};
+		}
+	}
+
+	@Configuration
+	static class OAuth2ClientWebMvcSecurityConfiguration implements WebMvcConfigurer {
+		@Autowired(required = false)
```

Comment:
I'm wondering if there is a better way to configure this? I don't like the fact that I need to add `@Autowired(required = false)` for both `ClientRegistrationRepository` and `OAuth2AuthorizedClientService` here but if I don't than tests will fail given that the conditions of Spring Web Mvc and OAuth 2 Client will always be in the class path during the build process.

Reply:
What don't you like about it? You should also use `ObjectProvider<ClientRegistrationRepository>` if you wanted. Does that help?

Reply:
`OAuth2ImportSelector` (via `@EnableWebSecurity`) will always import `OAuth2ClientConfiguration` because `oauth2-client` is on the class path...when running the build.  So both `ClientRegistrationRepository` and `OAuth2AuthorizedClientService` need to have `@Autowired(required = false)` or other tests that `@EnableWebSecurity` will fail with `BeanException`.

If `OAuth2ClientConfiguration` is imported than the assumption is that `ClientRegistrationRepository` and `OAuth2AuthorizedClientService` must be registered or else a `BeanException` will happen. So adding `@Autowired(required = false)` is a workaround (that I don't like) so those other tests don't fail during the build process.